Refactor buienradar sensor to reduce complexity and address code smells

zihanKuang opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

zihanKuang committed:
add-async back in config-flow file

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
refactor '_load_data' function to reduce cognitive

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
remove async

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

Bi9River commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -28,7 +28,7 @@
 SUPPORTED_COUNTRY_CODES = ["NL", "BE"]
 
 
-async def async_setup_entry(
```
Please don't remove the async keyword since there might be unexpected behavior involved.

Bi9River commented on homeassistant/components/buienradar/sensor.py:
```diff
@@ -60,6 +60,16 @@
 _LOGGER = logging.getLogger(__name__)
 
+# Icon constants
+ICON_GAUGE = "mdi:gauge"
```
Consts should be put in const.py

Bi9River commented on homeassistant/components/buienradar/sensor.py:
```diff
@@ -793,143 +793,142 @@ def _update(self):
+        if sensor_type.endswith(("_1d", "_2d", "_3d", "_4d", "_5d")):
+            fcday = int(sensor_type[-2]) - 1
+            if sensor_type.startswith((SYMBOL, CONDITION)):
+                value_updated = self._update_forecast_condition(
```
Using variable name like is_value_updated to indicate a bool variable would be more clear.

zihanKuang committed:
Revert "refactor '_load_data' function to reduce cognitive"

This reverts commit 6dc5412216280001a4ad6d0c129b202e6084475b.

zihanKuang committed:
move consts from sensor.py to const.py

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
refactor convert_windspeed_kph and convert_visibility_km

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
change if to elif

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
refactor _update_forecast_condition function

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
refactor _update_precipitation_forecast function

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
reduce depulicate code in _update_current_condition and _update_forecast_condition

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
fix some bugs in _load_data

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
fix bugs in _load_data

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
add async back

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang commented:
<img width="1041" height="471" alt="image" src="https://github.com/user-attachments/assets/8ffa3d2b-6f1f-4df4-bf15-38c1f40bf4fd" />

Bi9River reviewed (dismissed):
- Would be better if newly added functions have some short but clear comments
- IMHO the `fcday` shouldn't be updated using the string value. Instead, maybe we can have a separated function that can be used like this:

```
fcday = some_func(sensor_type)
```

So that it would be clear even if the sensor type changes or we need to modify the fcday calculation method.

Others looks good to me.

zihanKuang committed:
seperate _get_fcday_from_sensor for clean code

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
Add proper comments

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
remove Redundant try-catch

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang committed:
fix-bug:convert_visibility

Signed-off-by: Zihan Kuang <Zihan_kuang@outlook.com>

zihanKuang commented:
Update:

1. Added detailed inline comments to improve code readability and clarify logic flow.
2. Extracted the fcday determination logic into a helper function _get_fcday_from_sensor() to enhance modularity and extensibility.
3. Fixed a previous issue where visibility values in forecast data were not being converted from meters to kilometers.

<img width="1659" height="462" alt="image" src="https://github.com/user-attachments/assets/3d00ef8a-e413-4219-884a-9947ac7af46b" />
